### PR TITLE
Add Covariance Indicator

### PR DESCRIPTION
**Pull request by stevespringer**
Add indicator for covariance.

#### Description
Implement covariance indicator taking a target and reference symbol, and a look-back period.

_Note on code reuse:_ I was pointed towards the RateOfChange indicator as an example of reducing duplication, but I couldn't see way to do this without either creating a new base class (similar to WindowIndicator but with a reference window) or introducing a hacky way to hold onto the window passed into `ComputeNextValue()` by the base. I have therefore left the code as-is for now, but can will make changes if requested in review.

#### Related Issue
Closes https://github.com/QuantConnect/Lean/issues/6982

#### Requires Documentation Change

I think I'll need some input on what text to use for the summary of the new class `QuantConnect.Indicators.Covariance`. Not sure where these are sourced from as I'm a first-time contributor.

#### How Has This Been Tested?
Added unit tests, including comparison with reference data generated via script in https://github.com/QuantConnect/Lean/issues/6982#issuecomment-1777089341

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [ X] My code follows the code style of this project.
- [ X] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ X] I have added tests to cover my changes. 
- [ X] All new and existing tests passed.
- [ X] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Indicators/Covariance.cs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+            else if (inputSymbol == _referenceSymbol)
+            {
+                _referenceDataPoints.Add(input.Value);
+            }
+            else
+            {
+                throw new ArgumentException("The given symbol was not target or reference symbol");
+            }
+
+            if (_targetDataPoints.Samples == _referenceDataPoints.Samples && _referenceDataPoints.Count > 1)
```
Comment (Martin-Molinero) [inline Indicators/Covariance.cs:142]: Inside `ComputeCovariance` it's also checking for `IsReady`, should only have a single check for this 👍

File: Indicators/Covariance.cs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        private void ComputeCovariance()
+        {
+
+            if (!IsReady)
+            {
+                _covariance = 0m;
+                return;
+            }
+
+            decimal mean1 = 0;
```
Comment (Martin-Molinero) [inline Indicators/Covariance.cs:164]: In the lean code base we usually user var cause it's cleaner and try to use description names, for example `decimal mean1 = 0;` -> `var targetMean = 0m;`

File: Indicators/Covariance.cs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            for (int i = 0; i < windowSize; i++)
+            {
+                mean1 += _targetDataPoints[i] / windowSize;
+                mean2 += _referenceDataPoints[i] / windowSize;
+            }
+
+            for (int i = 0; i < windowSize; i++)
+            {
+                sumProduct += (_targetDataPoints[i] - mean1) * (_referenceDataPoints[i] - mean2);
+            }
```
Comment (Martin-Molinero) [inline Indicators/Covariance.cs:178]: Take a look at the beta indicator which is calculating the covariance internally, as pointed by potential solution of the issue, I would suggest extracting that logic into this new covariance implementation and make the beta indicator the first user of it 👍

File: Indicators/Covariance.cs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        public Covariance(string name, Symbol targetSymbol, Symbol referenceSymbol, int period)
+            : base(name)
+        {
+            // Assert the period is greater than two, otherwise the covariance can not be computed
+            if (period < 2)
+            {
+                throw new ArgumentException($"Period parameter for Covariance indicator must be greater than 2 but was {period}");
+            }
+
+            WarmUpPeriod = period + 1;
```
Comment (Martin-Molinero) [inline Indicators/Covariance.cs:79]: Mind explaining why it requires 1 extra value than the requested period?

File: Tests/TestData/spy_qqq_cov.csv [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+SPY,20200324,2349900,2441000,2338000,2431500,214693586,
+QQQ,20200324,1798800,1838300,1779500,1836600,86434644,25484181906.50064
+SPY,20200325,2450000,2563500,2397600,2467900,278538590,
+QQQ,20200325,1844900,1895100,1799800,1823000,82965783,25650271902.86495
+SPY,20200326,2492600,2628000,2490500,2612000,224153673,
+QQQ,20200326,1839300,1927200,1835600,1919000,74995622,25571440846.139008
+SPY,20200327,2532200,2608000,2510500,2534200,193638954,
+QQQ,20200327,1869000,1905100,1846600,1853000,54182139,25638656906.974693
+SPY,20200330,2556800,2623900,2535300,2616500,141153243,
+QQQ,20200330,1870600,1924900,1863500,1920400,46667707,25591109173.308296
```
Comment (Martin-Molinero) [inline Tests/TestData/spy_qqq_cov.csv:123]: These covariance values do not seem accurate? `25591109173.308296` the order of magnitude seems off. Believe the mean formular is off

File: Indicators/Covariance.cs [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+ * limitations under the License.
+*/
+
+using System;
+using QuantConnect.Data.Market;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// TODO: figure out what to write here
```
Comment (Martin-Molinero) [inline Indicators/Covariance.cs:22]: Should describe what this covariance indicator is, what's it calculating/showing/etc, you can see other indicators as an example.

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Hey @stevespringer! Thank you for the contribution. Leaving a few requests

File: Indicators/Covariance.cs (hunk 4 above)
Comment (stevespringer) [reply Indicators/Covariance.cs:79]: Sorry, I was taking the lead from another indicator without considering if it's relevant here. I'll remove the plus 1

File: Tests/TestData/spy_qqq_cov.csv (hunk 5 above)
Comment (Martin-Molinero) [reply Tests/TestData/spy_qqq_cov.csv:123]: See my other comments https://github.com/QuantConnect/Lean/pull/7843#discussion_r1520050464

**Comment by Martin-Molinero** [comment]
Closing for now, please reopen when review addressed @stevespringer

**Pull request closed without merging**